pyscf/solvent/hessian/pcm.py

```diff
@@ -0,0 +1,229 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+'''
+Gradient of PCM family solvent model, copied from GPU4PyCF with modifications
+'''
+# pylint: disable=C0103
+
+import numpy
+from pyscf import lib, gto
+from pyscf import scf
+from pyscf.solvent.pcm import PI
+from pyscf.solvent.grad.pcm import grad_qv, grad_solver, grad_nuc
+from pyscf.lib import logger
+
+def hess_nuc(pcmobj):
+    if not pcmobj._intermediates:
+        pcmobj.build()
+    mol = pcmobj.mol
+    q_sym        = pcmobj._intermediates['q_sym']
+    gridslice    = pcmobj.surface['gslice_by_atom']
+    grid_coords  = pcmobj.surface['grid_coords']
+    exponents    = pcmobj.surface['charge_exp']
+
+    atom_coords = mol.atom_coords(unit='B')
+    atom_charges = numpy.asarray(mol.atom_charges(), dtype=numpy.float64)
+    fakemol_nuc = gto.fakemol_for_charges(atom_coords)
+    fakemol = gto.fakemol_for_charges(grid_coords, expnt=exponents**2)
+
+    # nuclei potential response
+    int2c2e_ip1ip2 = mol._add_suffix('int2c2e_ip1ip2')
+    v_ng_ip1ip2 = gto.mole.intor_cross(int2c2e_ip1ip2, fakemol_nuc, fakemol).reshape([3,3,mol.natm,-1])
+    dv_g = numpy.einsum('n,xyng->ngxy', atom_charges, v_ng_ip1ip2)
+    dv_g = numpy.einsum('ngxy,g->ngxy', dv_g, q_sym)
+
+    de = numpy.zeros([mol.natm, mol.natm, 3, 3])
+    for ia in range(mol.natm):
+        p0, p1 = gridslice[ia]
+        de_tmp = numpy.sum(dv_g[:,p0:p1], axis=1)
+        de[:,ia] -= de_tmp
+        #de[ia,:] -= de_tmp.transpose([0,2,1])
+
+
+    int2c2e_ip1ip2 = mol._add_suffix('int2c2e_ip1ip2')
+    v_ng_ip1ip2 = gto.mole.intor_cross(int2c2e_ip1ip2, fakemol, fakemol_nuc).reshape([3,3,-1,mol.natm])
+    dv_g = numpy.einsum('n,xygn->gnxy', atom_charges, v_ng_ip1ip2)
+    dv_g = numpy.einsum('gnxy,g->gnxy', dv_g, q_sym)
+
+    for ia in range(mol.natm):
+        p0, p1 = gridslice[ia]
+        de_tmp = numpy.sum(dv_g[p0:p1], axis=0)
+        de[ia,:] -= de_tmp
+        #de[ia,:] -= de_tmp.transpose([0,2,1])
+
+    int2c2e_ipip1 = mol._add_suffix('int2c2e_ipip1')
+    v_ng_ipip1 = gto.mole.intor_cross(int2c2e_ipip1, fakemol_nuc, fakemol).reshape([3,3,mol.natm,-1])
+    dv_g = numpy.einsum('g,xyng->nxy', q_sym, v_ng_ipip1)
+    for ia in range(mol.natm):
+        de[ia,ia] -= dv_g[ia] * atom_charges[ia]
+
+    v_ng_ipip1 = gto.mole.intor_cross(int2c2e_ipip1, fakemol, fakemol_nuc).reshape([3,3,-1,mol.natm])
+    dv_g = numpy.einsum('n,xygn->gxy', atom_charges, v_ng_ipip1)
+    dv_g = numpy.einsum('g,gxy->gxy', q_sym, dv_g)
+    for ia in range(mol.natm):
+        p0, p1 = gridslice[ia]
+        de[ia,ia] -= numpy.sum(dv_g[p0:p1], axis=0)
+
+    return de
+
```

Comment:
For finite difference hessians, it would be better to print a warning saying this is computed with finite difference.

Reply:
Added